src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1599,19 +1603,27 @@ private SolutionCompilationState ComputeFrozenSnapshot(CancellationToken cancell
                 if (oldTracker == newTracker)
                     continue;
 
-                Contract.ThrowIfFalse(newIdToProjectStateMapBuilder.ContainsKey(projectId));
-
                 var newProjectState = newTracker.ProjectState;
 
-                newIdToProjectStateMapBuilder[projectId] = newProjectState;
+                newProjectStatesBuilder[i] = newProjectState;
                 newIdToTrackerMapBuilder[projectId] = newTracker;
+                projectStateChanged = true;
             }
 
             // Attempt to update _projectIdToTrackerMap to include all the newly created compilation trackers. If another thread has updated
             // it since we captured it, then we'll need to loop again to ensure we've operated on the latest compilation trackers.
             var updatedIdToTrackerMap = updatedIdToTrackerMapBuilder.ToImmutable();
             if (originalProjectIdToTrackerMap == RoslynImmutableInterlocked.InterlockedCompareExchange(ref _projectIdToTrackerMap, updatedIdToTrackerMap, originalProjectIdToTrackerMap))
-                return (newIdToProjectStateMapBuilder.ToImmutable(), newIdToTrackerMapBuilder.ToImmutable());
+            {
+                var newProjectStates = this.SolutionState.ProjectStates;
+                if (projectStateChanged)
+                {
+                    newProjectStatesBuilder.Sort(SolutionState.CompareProjectStates);
```

Comment:
If we enumerated in order, didn't we keep the ordering? So how could this change?

Reply:
There's a lot going on in here with trackers and whatnot to create the new project state. If we're certain that their project id is the same, then this can be simplified. I'll go ahead and make that change and test it out.